File: CSharp/SpaceVIL/SpaceVIL/Items/VisualItem.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Threading;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// Common type of events. No arguments.
    /// </summary>
    public delegate void EventCommonMethod();
    /// <summary>
    /// Common type of events with one argument: sender.
    /// </summary>
    /// <param name="sender">Sender as SpaceVIL.Core.IItem.</param>
    public delegate void EventCommonMethodState(IItem sender);
    /// <summary>
    /// Mouse type of event with two arguments: sender, args.
    /// </summary>
    /// <param name="sender">Sender as SpaceVIL.Core.IItem.</param>
    /// <param name="args">Mouse arguments as SpaceVIL.Core.MouseArgs.</param>
    public delegate void EventMouseMethodState(IItem sender, MouseArgs args);
    /// <summary>
    /// Key type of event with two arguments: sender, args.
    /// </summary>
    /// <param name="sender">Sender as SpaceVIL.Core.IItem.</param>
    /// <param name="args">Key arguments as SpaceVIL.Core.KeyArgs.</param>
    public delegate void EventKeyMethodState(IItem sender, KeyArgs args);
    /// <summary>
    /// Text input type of event with two arguments: sender, args.
    /// </summary>
    /// <param name="sender">Sender as SpaceVIL.Core.IItem.</param>
    /// <param name="args">Text input arguments as SpaceVIL.Core.TextInputArgs.</param>
    public delegate void EventInputTextMethodState(IItem sender, TextInputArgs args);

    /// <summary>
    /// Drop type of event with two arguments: sender, args.
    /// </summary>
    /// <param name="sender">Sender as SpaceVIL.Core.IItem.</param>
    /// <param name="args">Drop arguments as SpaceVIL.Core.DropArgs.</param>
    public delegate void EventWindowDropMethod(IItem sender, DropArgs args);

    internal class VisualItem : BaseItem
    {
        private Object _locker = new Object();
        internal Prototype prototype;

        private String _tooltip = String.Empty;
        internal String GetToolTip()
        {
            return _tooltip;
        }
        internal void SetToolTip(String text)
        {
            _tooltip = text;
        }
        //container
        private Spacing _spacing = new Spacing();
        internal Spacing GetSpacing()
        {
            return _spacing;
        }
        internal void SetSpacing(Spacing spacing)
        {
            _spacing = spacing;
            UpdateGeometry();
            BaseItemStatics.UpdateAllLayout(this);
        }
        internal void SetSpacing(int horizontal = 0, int vertical = 0)
        {
            SetSpacing(new Spacing(horizontal, vertical));
        }
        private Indents _padding = new Indents();
        internal Indents GetPadding()
        {
            return _padding;
        }
        internal void SetPadding(Indents padding)
        {
            _padding = padding;
            UpdateGeometry();
            BaseItemStatics.UpdateAllLayout(this);
        }
        internal void SetPadding(int left = 0, int top = 0, int right = 0, int bottom = 0)
        {
            SetPadding(new Indents(left, top, right, bottom));
        }
        internal EventManager eventManager = null;
        // private List<IBaseItem> _content = new List<IBaseItem>();
        private LinkedHashSet<IBaseItem> _content = new LinkedHashSet<IBaseItem>();
        internal virtual List<IBaseItem> GetItems()
        {
            Monitor.Enter(_locker);
            try
            {
                return new List<IBaseItem>(_content);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
                return null;
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }
        internal void SetContent(List<IBaseItem> content)
        {
            Monitor.Enter(_locker);
            try
            {
                // _content = content;
                _content = new LinkedHashSet<IBaseItem>(content);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }

        private void CastAndRemove(IBaseItem item)
        {
            Prototype prototype = item as Prototype;
            if (prototype != null)
            {
                prototype.GetCore().RemoveItemFromListeners();
                prototype.FreeEvents();
            }
            else
                (item as BaseItem).RemoveItemFromListeners();
        }

        internal override void RemoveItemFromListeners()
        {
            Prototype parent = GetParent();
            parent.RemoveEventListener(GeometryEventType.ResizeWidth, this.prototype);
            parent.RemoveEventListener(GeometryEventType.ResizeHeight, this.prototype);
            parent.RemoveEventListener(GeometryEventType.MovedX, this.prototype);
            parent.RemoveEventListener(GeometryEventType.MovedY, this.prototype);
        }

        internal virtual void AddItem(IBaseItem item)
        {
            Monitor.Enter(_locker);
            try
            {
                if (item == null)
                {
                    Console.WriteLine("SpaceVILWarning: Trying to add null item!");
                    return;
                }
                if (item.Equals(this))
                {
                    Console.WriteLine("SpaceVILWarning: Trying to add current item in himself.");
                    return;
                }
                item.SetHandler(GetHandler());
                AddChildren(item);
                _content.Add(item);
                ItemsLayoutBox.AddItem(GetHandler(), item, LayoutType.Static);

                //needs to force update all attributes
                BaseItemStatics.CastToUpdateGeometry(item);
                item.InitElements();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Method - AddItem: " + ((item == null) ? "item is null" : item.GetItemName()));
                Console.WriteLine(ex.StackTrace);
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }
        internal virtual void InsertItem(IBaseItem item, Int32 index)
        {
            Monitor.Enter(_locker);
            try
            {
                if (item == null)
                {
                    Console.WriteLine("Trying to insert null item");
                    return;
                }
                if (index < 0)
                {
                    Console.WriteLine("Invalid index");
                    return;
                }
                if (item.Equals(this))
                {
                    Console.WriteLine("Trying to add current item in himself.");
                    return;
                }
                item.SetHandler(GetHandler());
                AddChildren(item);
                if (index > _content.Count)
                    _content.Add(item);
                else
                {
                    List<IBaseItem> list = new List<IBaseItem>(_content);
                    list.Insert(index, item);
                    _content = new LinkedHashSet<IBaseItem>(list);
                }
                ItemsLayoutBox.AddItem(GetHandler(), item, LayoutType.Static);
                //needs to force update all attributes
                BaseItemStatics.CastToUpdateGeometry(item);
                item.InitElements();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Method - InsertItem: " + ((item == null) ? "item is null" : item.GetItemName()));
                Console.WriteLine(ex.StackTrace);
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }

        internal void CascadeRemoving(IBaseItem item, LayoutType type)
        {
            Prototype container = item as Prototype;
            if (container != null)
            {
                List<IBaseItem> tmp = container.GetItems();
                while (tmp.Count > 0)
                {
                    IBaseItem child = tmp.ElementAt(0);
                    container.RemoveItem(child);
                    tmp.Remove(child);
                }
            }
        }

        internal virtual bool RemoveItem(IBaseItem item)
        {
            Monitor.Enter(_locker);
            try
            {
                if (!_content.Contains(item))
                    return false;
                //reset focus
                Prototype tmp = item as Prototype;
                if (tmp != null)
                {
                    if (tmp.IsFocused())
                        GetHandler().ResetItems();
                }

                LayoutType type;
                if (item is IFloating)
                {
                    CascadeRemoving(item, LayoutType.Floating);
                    type = LayoutType.Floating;
                }
                else
                {
                    CascadeRemoving(item, LayoutType.Static);
                    type = LayoutType.Static;
                }

                //removing
                CastAndRemove(item);
                bool contentRemove = _content.Remove(item);
                bool layoutBoxRemove = ItemsLayoutBox.RemoveItem(GetHandler(), item, type);
                item.SetParent(null);
                item.Release();
                return (contentRemove && layoutBoxRemove);
            }
            catch (Exception ex)
            {
                Console.WriteLine((item == null) ? "item is null" : item.GetItemName() + "\n" + ex.ToString());
                Console.WriteLine(ex.StackTrace);
                return false;
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }

        internal virtual void Clear()
        {
            Monitor.Enter(_locker);
            try
            {
                while (_content.Count > 0)
                    RemoveItem(_content.First());
            }
            catch (Exception ex)
            {
                Console.WriteLine("Method - Clear");
                Console.WriteLine(ex.StackTrace);
            }
            finally
            {
                Monitor.Exit(_locker);
            }
        }

        internal override void AddEventListener(GeometryEventType type, IBaseItem listener)
        {
            eventManager.Subscribe(type, listener);
        }
        internal override void RemoveEventListener(GeometryEventType type, IBaseItem listener)
        {
            eventManager.Unsubscribe(type, listener);
        }

        //item
        internal Border border = new Border();
        internal void SetBorderDirect(Border border)
        {
            this.border = border;
        }

        internal Border GetBorderDirect()
        {
            return border;
        }

        internal void SetBorder(Border border)
        {
            this.border = border;
            GetState(ItemStateType.Base).Border = this.border;
            UpdateState();
        }

        internal void SetBorderFill(Color fill)
        {
            border.SetFill(fill);
            GetState(ItemStateType.Base).Border.SetFill(fill);
            UpdateState();
        }
        internal virtual void SetBorderFill(int r, int g, int b)
        {
            SetBorderFill(GraphicsMathService.ColorTransform(r, g, b));
        }
        internal virtual void SetBorderFill(int r, int g, int b, int a)
        {
            SetBorderFill(GraphicsMathService.ColorTransform(r, g, b, a));
        }
        internal virtual void SetBorderFill(float r, float g, float b)
        {
            SetBorderFill(GraphicsMathService.ColorTransform(r, g, b));
        }
        internal virtual void SetBorderFill(float r, float g, float b, float a)
        {
            SetBorderFill(GraphicsMathService.ColorTransform(r, g, b, a));
        }

        internal void SetBorderRadius(CornerRadius radius)
        {
            border.SetRadius(radius);
            GetState(ItemStateType.Base).Border.SetRadius(radius);
            UpdateState();
        }
        internal void SetBorderThickness(int thickness)
        {
            border.SetThickness(thickness);
            GetState(ItemStateType.Base).Border.SetThickness(thickness);
            UpdateState();
        }
        internal CornerRadius GetBorderRadius()
        {
            return border.GetRadius();
        }
        internal int GetBorderThickness()
        {
            return border.GetThickness();
        }
        internal Color GetBorderFill()
        {
            return border.GetFill();
        }


        internal void SetBorder(Color fill, CornerRadius radius, int thickness)
        {
            border = new Border(fill, radius, thickness);
            GetState(ItemStateType.Base).Border = border;
            UpdateState();
        }

        internal Dictionary<ItemStateType, ItemState> states = new Dictionary<ItemStateType, ItemState>();
        internal ItemStateType _state = ItemStateType.Base;
        internal ItemStateType GetCurrentStateType()
        {
            return _state;
        }
        internal void SetState(ItemStateType state)
        {
            _state = state;
            UpdateState();
        }

        internal VisualItem(
        int xpos = 0,
        int ypos = 0,
        int width = 0,
        int height = 0,
        string name = "VisualItem")
        {
            states.Add(ItemStateType.Base, new ItemState()
            {
                Value = true,
                Background = GetBackground(),
            });

            //common default prop
            eventManager = new EventManager();
            SetItemName(name);
        }

        //overrides
        public override void SetWidth(int width)
        {
            int value = width - GetWidth();
            if (value != 0)
            {
                base.SetWidth(width);
                BaseItemStatics.UpdateHLayout(this);
                eventManager.NotifyListeners(GeometryEventType.ResizeWidth, value);
            }
        }
        public override void SetHeight(int height)
        {
            int value = height - GetHeight();
            if (value != 0)
            {
                base.SetHeight(height);
                BaseItemStatics.UpdateVLayout(this);
                eventManager.NotifyListeners(GeometryEventType.ResizeHeight, value);
            }
        }
        internal void SetPosition(int x, int y)
        {
            this.SetX(x);
            this.SetY(y);
        }
        public override void SetX(int x)
        {
            int value = x - GetX();
            if (value != 0)
            {
                base.SetX(x);
                eventManager.NotifyListeners(GeometryEventType.MovedX, value);
            }
        }
        public override void SetY(int y)
        {
            int value = y - GetY();
            if (value != 0)
            {
                base.SetY(y);
                eventManager.NotifyListeners(GeometryEventType.MovedY, value);
            }
        }

        // common properties
        private InputEventType _blockedEvents = 0x00;

        internal bool IsPassEvents()
        {
            if (_blockedEvents == 0x00)
                return true;
            return false;
        }

        internal bool IsPassEvents(InputEventType e)
        {
            if (_blockedEvents.HasFlag(e))
                return false;
            return true;
        }

        internal List<InputEventType> GetBlockedEvents()
        {
            List<InputEventType> list = new List<InputEventType>();
            foreach (InputEventType type in Enum.GetValues(typeof(InputEventType)))
            {
                if (_blockedEvents.HasFlag(type))
                {
                    list.Add(type);
                }
            }

            return list;
        }

        internal List<InputEventType> GetPassEvents()
        {
            List<InputEventType> list = new List<InputEventType>();
            foreach (InputEventType type in Enum.GetValues(typeof(InputEventType)))
            {
                if (!_blockedEvents.HasFlag(type))
                {
                    list.Add(type);
                }
            }

            return list;
        }

        internal void SetPassEvents(bool value)
        {
            if (!value)
            {
                foreach (InputEventType e in Enum.GetValues(typeof(InputEventType)))
                {
                    _blockedEvents |= e;
                }
            }
            else
            {
                _blockedEvents = 0x00;
            }
        }

        internal void SetPassEvents(bool value, InputEventType e)
        {
            if (!value)
            {
                _blockedEvents |= e;
            }
            else
            {
                if (_blockedEvents.HasFlag(e))
                    _blockedEvents &= ~e;
            }
        }

        internal void SetPassEvents(bool value, params InputEventType[] e)
        {
            foreach (var item in e)
            {
                SetPassEvents(value, item);
            }
        }

        private bool _disabled;

        internal bool IsDisabled()
        {
            return _disabled;
        }

        internal void SetDisabled(bool value)
        {
            if (_disabled == value)
                return;
            _disabled = value;
            UpdateState();
        }

        private bool _hover;

        internal bool IsMouseHover()
        {
            return _hover;
        }

        internal void SetMouseHover(bool value)
        {
            if (_hover == value)
                return;
            _hover = value;
            UpdateState();
        }

        private bool _pressed;
        internal bool IsMousePressed()
        {
            return _pressed;
        }

        internal void SetMousePressed(bool value)
        {
            if (_pressed == value)
                return;
            _pressed = value;
            UpdateState();
        }

        private bool _focused;
        internal bool IsFocused()
        {
            return _focused;
        }

        internal void SetFocused(bool value)
        {
            if (_focused == value)
                return;
            _focused = value;
            UpdateState();
        }

        private bool _focusable = true;
        internal bool IsFocusable()
        {
            return _focusable;
        }
        internal void SetFocusable()
        {
            //foreach inner item focusable value set?
        }

        //common methods
        internal void AddItemState(ItemStateType type, ItemState state)
        {
            if (states.ContainsKey(type))
            {
                state.Value = true;
                states[type] = state;
            }
            else
            {
                states.Add(type, state);
            }
        }
        internal void RemoveItemState(ItemStateType type)
        {
            if (type == ItemStateType.Base)
                return;
            if (states.ContainsKey(type))
                states.Remove(type);
        }
        internal void RemoveAllItemStates()
        {
            var itemsToRemove = states.Where(f => f.Key != ItemStateType.Base).ToArray();
            foreach (var item in itemsToRemove)
                states.Remove(item.Key);
        }
        internal ItemState GetState(ItemStateType type)
        {
            if (states.ContainsKey(type))
                return states[type];
            return null;
        }

        internal void SetBackgroundDirect(Color color)
        {
            base.SetBackground(color);
        }

        public override void SetBackground(Color color)
        {
            GetState(ItemStateType.Base).Background = color;
            UpdateState();
        }
        public override void SetBackground(int r, int g, int b)
        {
            base.SetBackground(r, g, b);
            GetState(ItemStateType.Base).Background = GetBackground();
            UpdateState();
        }
        public override void SetBackground(int r, int g, int b, int a)
        {
            base.SetBackground(r, g, b, a);
            GetState(ItemStateType.Base).Background = GetBackground();
            UpdateState();
        }
        public override void SetBackground(float r, float g, float b)
        {
            base.SetBackground(r, g, b);
            GetState(ItemStateType.Base).Background = GetBackground();
            UpdateState();
        }
        public override void SetBackground(float r, float g, float b, float a)
        {
            base.SetBackground(r, g, b, a);
            GetState(ItemStateType.Base).Background = GetBackground();
            UpdateState();
        }

        internal void UpdateState()
        {
            VisualItemStatics.UpdateState(this);
        }

        internal virtual bool GetHoverVerification(float xpos, float ypos)
        {
            return VisualItemStatics.GetHoverVerification(this, xpos, ypos);
        }

        Figure IsCustom = null;
        internal Figure IsCustomFigure()
        {
            return IsCustom;
        }
        internal void SetCustomFigure(Figure figure)
        {
            IsCustom = figure;
            ItemsRefreshManager.SetRefreshShape(this.prototype);
        }

        public override void MakeShape()
        {
            if (IsCustomFigure() != null)
            {
                SetTriangles(IsCustomFigure().GetFigure());
                if (GetState(ItemStateType.Base).Shape == null)
                    GetState(ItemStateType.Base).Shape = IsCustomFigure();

                if (!IsCustomFigure().IsFixed())
                    SetTriangles(UpdateShape());
            }
            else
            {
                SetTriangles(GraphicsMathService.GetRoundSquare(GetBorderRadius(), GetWidth(), GetHeight(), 0, 0));
            }
        }

        public override void SetStyle(Style style)
        {
            VisualItemStatics.SetStyle(this, style);
        }

        public override Style GetCoreStyle()
        {
            return VisualItemStatics.ExtractCoreStyle(this);
        }

        public override void InitElements() { }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/VisualItem.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.*;
import com.spvessel.spacevil.Decorations.*;
import com.spvessel.spacevil.Flags.*;

import java.awt.Color;
import java.util.List;
import java.util.LinkedList;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.*;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashSet;

final class VisualItem extends BaseItem {

    private Lock locker = new ReentrantLock();

    Prototype prototype;

    VisualItem() {
        this("VisualItem_");
    }

    VisualItem(String name) {
        ItemState base = new ItemState();
        base.background = getBackground();
        states.put(ItemStateType.Base, base);

        // common default prop
        eventManager = new EventManager();
        setItemName(name);
    }

    // overrides
    @Override
    public void setWidth(int width) {
        int value = width - getWidth();
        if (value != 0) {
            super.setWidth(width);
            BaseItemStatics.updateHLayout(this);
            eventManager.notifyListeners(GeometryEventType.ResizeWidth, value);
        }
    }

    @Override
    public void setHeight(int height) {
        int value = height - getHeight();
        if (value != 0) {
            super.setHeight(height);
            BaseItemStatics.updateVLayout(this);
            eventManager.notifyListeners(GeometryEventType.ResizeHeight, value);
        }
    }

    void setPosition(int x, int y) {
        this.setX(x);
        this.setY(y);
    }

    @Override
    public void setX(int x) {
        int value = x - getX();
        if (value != 0) {
            super.setX(x);
            eventManager.notifyListeners(GeometryEventType.MovedX, value);
        }
    }

    @Override
    public void setY(int y) {
        int value = y - getY();
        if (value != 0) {
            super.setY(y);
            eventManager.notifyListeners(GeometryEventType.MovedY, value);
        }
    }

    // item
    Border border = new Border();

    void setBorderDirect(Border border) {
        this.border = border;
    }

    Border getBorder() {
        return border;
    }

    void setBorder(Border border) {
        this.border = border;
        getState(ItemStateType.Base).border = border;
        updateState();
    }

    void setBorderFill(Color color) {
        border.setColor(color);
        getState(ItemStateType.Base).border.setColor(color);
        updateState();
    }

    void setBorderFill(int r, int g, int b) {
        setBorderFill(GraphicsMathService.colorTransform(r, g, b));
    }

    void setBorderFill(int r, int g, int b, int a) {
        setBorderFill(GraphicsMathService.colorTransform(r, g, b, a));
    }

    void setBorderFill(float r, float g, float b) {
        setBorderFill(GraphicsMathService.colorTransform(r, g, b));
    }

    void setBorderFill(float r, float g, float b, float a) {
        setBorderFill(GraphicsMathService.colorTransform(r, g, b, a));
    }

    void setBorderRadius(CornerRadius radius) {
        border.setRadius(radius);
        getState(ItemStateType.Base).border.setRadius(radius);
        updateState();
    }

    void setBorderThickness(int thickness) {
        border.setThickness(thickness);
        getState(ItemStateType.Base).border.setThickness(thickness);
        updateState();
    }

    CornerRadius getBorderRadius() {
        return border.getRadius();
    }

    int getBorderThickness() {
        return border.getThickness();
    }

    Color getBorderFill() {
        return border.getColor();
    }

    void SetBorder(Color color, CornerRadius radius, int thickness) {
        border.setColor(color);
        border.setRadius(radius);
        border.setThickness(thickness);
        getState(ItemStateType.Base).border = border;
        updateState();
    }

    Map<ItemStateType, ItemState> states = new HashMap<>();
    ItemStateType _state = ItemStateType.Base;

    void setState(ItemStateType state) {
        _state = state;
        updateState();
    }

    ItemStateType getCurrentStateType() {
        return _state;
    }

    private String _tooltip = "";

    String getToolTip() {
        return _tooltip;
    }

    void setToolTip(String text) {
        _tooltip = text;
    }

    // container
    private Spacing _spacing = new Spacing();

    Spacing getSpacing() {
        return _spacing;
    }

    void setSpacing(Spacing spacing) {
        _spacing = spacing;
        updateGeometry();
        BaseItemStatics.updateAllLayout(this);
    }

    void setSpacing(int horizontal, int vertical) {
        setSpacing(new Spacing(horizontal, vertical));
    }

    private Indents _padding = new Indents();

    Indents getPadding() {
        return _padding;
    }

    void setPadding(Indents padding) {
        _padding = padding;
        updateGeometry();
        BaseItemStatics.updateAllLayout(this);
    }

    void setPadding(int left, int top, int right, int bottom) {
        setPadding(new Indents(left, top, right, bottom));
    }

    EventManager eventManager = null;
    private Set<IBaseItem> _content = new LinkedHashSet<>();

    List<IBaseItem> getItems() {
        locker.lock();
        try {
            return new LinkedList<IBaseItem>(_content);
        } catch (Exception ex) {
            ex.printStackTrace();
            return null;
        } finally {
            locker.unlock();
        }
    }

    void setContent(List<IBaseItem> content) {
        locker.lock();
        try {
            _content = new LinkedHashSet<>(content);
        } catch (Exception ex) {
            ex.printStackTrace();
        } finally {
            locker.unlock();
        }
    }

    private void castAndRemove(IBaseItem item) {
        if (item instanceof Prototype) {
            Prototype prototype = ((Prototype) item);
            prototype.getCore().removeItemFromListeners();
            prototype.freeEvents();
        } else {
            ((BaseItem) item).removeItemFromListeners();
        }
    }

    @Override
    void removeItemFromListeners() {
        Prototype parent = getParent();
        parent.removeEventListener(GeometryEventType.ResizeWidth, this.prototype);
        parent.removeEventListener(GeometryEventType.ResizeHeight, this.prototype);
        parent.removeEventListener(GeometryEventType.MovedX, this.prototype);
        parent.removeEventListener(GeometryEventType.MovedY, this.prototype);
    }

    void addItem(IBaseItem item) {
        locker.lock();
        try {
            if (item == null) {
                System.out.println("Trying to add null item");
                return;
            }
            if (item.equals(this)) {
                System.out.println("Trying to add current item in himself.");
                return;
            }
            item.setHandler(getHandler());
            addChildren(item);
            _content.add(item);
            ItemsLayoutBox.addItem(getHandler(), item, LayoutType.Static);
            // needs to force update all attributes
            BaseItemStatics.castToUpdateGeometry(item);
            item.initElements();
        } catch (Exception ex) {
            System.out.println("Method - AddItem: " + ((item == null) ? "item is null" : item.getItemName()));
            ex.printStackTrace();
        } finally {
            locker.unlock();
        }
    }

    void insertItem(IBaseItem item, int index) {
        locker.lock();
        try {
            if (item == null) {
                System.out.println("Trying to insert null item");
                return;
            }
            if (index < 0) {
                System.out.println("Invalid index");
                return;
            }
            if (item.equals(this)) {
                System.out.println("Trying to add current item in himself.");
                return;
            }
            item.setHandler(getHandler());

            addChildren(item);

            if (index > _content.size())
                _content.add(item);
            else {
                List<IBaseItem> list = new LinkedList<>(_content);
                list.add(index, item);
                _content = new LinkedHashSet<>(list);
            }

            try {
                ItemsLayoutBox.addItem(getHandler(), item, LayoutType.Static);
            } catch (Exception ex) {
                System.out.println(item.getItemName());
                throw ex;
            }

            // needs to force update all attributes
            BaseItemStatics.castToUpdateGeometry(item);
            item.initElements();

            // if (item instanceof VisualItem) {
            // ((VisualItem) item).updateState();
            // }
        } catch (Exception ex) {
            System.out.println("Method - InsertItem: " + ((item == null) ? "item is null" : item.getItemName()));
            ex.printStackTrace();
        } finally {
            locker.unlock();
        }
    }

    void cascadeRemoving(IBaseItem item, LayoutType type) {
        if (item instanceof Prototype)// и если это действительно контейнер
        {
            Prototype container = (Prototype) item;// предполагаю что элемент контейнер
            List<IBaseItem> tmp = container.getItems();
            while (tmp.size() > 0) {
                IBaseItem child = tmp.get(0);
                container.removeItem(child);
                tmp.remove(child);
            }
        }
    }

    boolean removeItem(IBaseItem item) {
        locker.lock();
        try {
            if (!_content.contains(item))
                return false;
            if (item instanceof Prototype) {
                Prototype tmp = ((Prototype) item);
                if (tmp.isFocused())
                    getHandler().resetItems();
            }

            LayoutType type;
            if (item instanceof IFloating) {
                cascadeRemoving(item, LayoutType.Floating);
                type = LayoutType.Floating;
            } else {
                cascadeRemoving(item, LayoutType.Static);
                type = LayoutType.Static;
            }

            // removing
            castAndRemove(item);
            boolean contentRemove = _content.remove(item);
            boolean layoutBoxRemove = ItemsLayoutBox.removeItem(getHandler(), item, type);
            item.setParent(null);
            item.release();
            return (contentRemove && layoutBoxRemove);
        } catch (Exception ex) {
            System.out.println("Method - RemoveItem: " + ((item == null) ? "item is null" : item.getItemName()));
            ex.printStackTrace();
            return false;
        } finally {
            locker.unlock();
        }
    }

    void clear() {
        locker.lock();
        try {
            while (!_content.isEmpty()) {
                removeItem(_content.iterator().next());
            }
        } catch (Exception ex) {
            System.out.println("Method - Clear");
            ex.printStackTrace();
        } finally {
            locker.unlock();
        }
    }

    @Override
    void addEventListener(GeometryEventType type, IBaseItem listener) {
        eventManager.subscribe(type, listener);
    }

    @Override
    void removeEventListener(GeometryEventType type, IBaseItem listener) {
        eventManager.unsubscribe(type, listener);
    }

    void addItemState(ItemStateType type, ItemState state) {
        if (states.containsKey(type)) {
            state.value = true;
            states.replace(type, state);
        } else {
            states.put(type, state);
        }
    }

    ItemState getState(ItemStateType type) {
        if (states.containsKey(type))
            return states.get(type);
        return null;
    }

    Map<ItemStateType, ItemState> getAllStates() {
        return states;
    }

    void removeItemState(ItemStateType type) {
        if (type == ItemStateType.Base)
            return;
        if (states.containsKey(type))
            states.remove(type);
    }

    void removeAllItemStates() {
        List<ItemStateType> itemsToRemove = states.entrySet().stream().filter(i -> i.getKey() != ItemStateType.Base)
                .map(Map.Entry::getKey).collect(Collectors.toList());
        for (ItemStateType item : itemsToRemove)
            states.remove(item);
    }

    final void setBackgroundDirect(Color color) {
        super.setBackground(color);
    }

    @Override
    public void setBackground(Color color) {
        getState(ItemStateType.Base).background = color;
        updateState();
    }

    @Override
    public void setBackground(int r, int g, int b) {
        super.setBackground(r, g, b);
        getState(ItemStateType.Base).background = getBackground();
        updateState();
    }

    @Override
    public void setBackground(int r, int g, int b, int a) {
        super.setBackground(r, g, b, a);
        getState(ItemStateType.Base).background = getBackground();
        updateState();
    }

    @Override
    public void setBackground(float r, float g, float b) {
        super.setBackground(r, g, b);
        getState(ItemStateType.Base).background = getBackground();
        updateState();
    }

    @Override
    public void setBackground(float r, float g, float b, float a) {
        super.setBackground(r, g, b, a);
        getState(ItemStateType.Base).background = getBackground();
        updateState();
    }

    // common properties
    private List<InputEventType> _blockedEvents = new LinkedList<>();

    boolean isPassEvents() {
        return (_blockedEvents.size() == 0);
    }

    List<InputEventType> getPassEvents() {
        List<InputEventType> result = Arrays.asList(InputEventType.values());
        return result.stream().filter(e -> !_blockedEvents.contains(e)).collect(Collectors.toList());
    }

    List<InputEventType> getBlockedEvents() {
        return _blockedEvents;
    }

    void setPassEvents(boolean value) {
        if (!value) {
            for (InputEventType e : Arrays.asList(InputEventType.values())) {
                _blockedEvents.add(e);
            }
        } else {
            _blockedEvents.clear();
        }
    }

    void setPassEvents(boolean value, InputEventType e) {
        if (!value) {
            if (!_blockedEvents.contains(e))
                _blockedEvents.add(e);
        } else {
            if (_blockedEvents.contains(e))
                _blockedEvents.remove(e);
        }
    }

    void setPassEvents(boolean value, List<InputEventType> event_set) {
        if (!value) {
            _blockedEvents = event_set;
        } else {
            for (InputEventType e : event_set) {
                if (_blockedEvents.contains(e))
                    _blockedEvents.remove(e);
            }
        }
    }

    private boolean _disabled;

    boolean isDisabled() {
        return _disabled;
    }

    void setDisabled(boolean value) {
        if (_disabled == value)
            return;
        _disabled = value;
        updateState();
    }

    private boolean _hover;

    boolean isMouseHover() {
        return _hover;
    }

    void setMouseHover(boolean value) {
        if (_hover == value)
            return;
        _hover = value;
        updateState();
    }

    private boolean _pressed;

    boolean isMousePressed() {
        return _pressed;
    }

    void setMousePressed(boolean value) {
        if (_pressed == value)
            return;
        _pressed = value;
        updateState();
    }

    private boolean _focused;
    // boolean isFocusable = true;

    boolean isFocused() {
        return _focused;
    }

    void setFocused(boolean value) {
        if (_focused == value)
            return;
        _focused = value;
        updateState();
    }

    // private boolean _focusable = true;

    // public boolean isFocusable() {
    // return _focusable;
    // }

    // public void setFocusable() {
    // // foreach inner item focusable value set?
    // }

    void updateState() {
        VisualItemStatics.updateState(this);
    }

    boolean getHoverVerification(float xpos, float ypos) {
        return VisualItemStatics.getHoverVerification(this, xpos, ypos);
    }

    private Figure _is_custom = null;

    final Figure isCustomFigure() {
        return _is_custom;
    }

    void setCustomFigure(Figure figure) {
        _is_custom = figure;
        ItemsRefreshManager.setRefreshShape(this.prototype);
    }

    @Override
    public void makeShape() {
        if (isCustomFigure() != null) {
            setTriangles(isCustomFigure().getFigure());
            if (getState(ItemStateType.Base).shape == null)
                getState(ItemStateType.Base).shape = isCustomFigure();

            if (!isCustomFigure().isFixed())
                setTriangles(updateShape());
        } else {
            setTriangles(GraphicsMathService.getRoundSquare(getBorderRadius(), getWidth(), getHeight(), 0, 0));
        }
    }

    @Override
    public void setStyle(Style style) {
        VisualItemStatics.setStyle(this, style);
    }

    @Override
    public Style getCoreStyle() {
        return VisualItemStatics.extractCoreStyle(this);
    }
}